### PR TITLE
Create postGenerateComparisonSchema event

### LINKED ISSUE #1406: doctrine:schema:validate returns DROP TABLE doctrine_migration_versions;
### BC Break Report

|    Q        |   A
|------------ | ------
| BC Break    | yes
| Version     | 3.0.0

#### Summary

When I updated doctrine/orm to version 3.0.0, my CI started to fail due to the doctrine:schema check:validate, which returned DROP TABLE doctrine_migration_versions;

I found that adding schema_filter can fix this problem: "~^(?!doctrine_migration_versions$)~" in the dbal configuration, indeed, then the error no longer appears, but after that my migrations stopped tracking the current state and think that at the moment no migration has been performed.

In version 3.0, the saveMode argument was removed in \Doctrine\ORM\Tools\SchemaTool::getUpdateSchemaSql and because of this, doctrine:schema:validate tracks deleted tables and returns sql for this action.
telegram-cloud-photo-size-2-5384566650814190356-y

#### Previous behavior

image

#### Current behavior

image

#### How to reproduce

1. composer require doctrine/doctrine-migrations-bundle
2. composer require doctrine/orm-pack
3. bin/console doctrine:migration:diff
4. bin/console doctrine:migration:migrate
5. bin/console doctrine:schema:validate -v

### PR DESCRIPTION
**Pull request by wmouwen**
Schemas generated from metadata cause the event `ToolEvents::postGenerateSchema` to be triggered in the method `SchemaTool::getSchemaFromMetadata`. It would be helpful to do the same for schemas generated for comparison in `SchemaTool::createSchemaForComparison`. This PR introduces the new event `ToolEvents::postGenerateComparisonSchema`. 

PR relates to an attempt to fix doctrine/migrations#1406. It is a requirement of doctrine/migrations#1418 and doctrine/DoctrineMigrationsBundle#529.

### CODE REVIEW

**Comment by greg0ire** [comment]
It seems there are CI jobs failing. Please take a look at [this guide](https://www.doctrine-project.org/contribute/#dealing-with-checks-and-tools) for more on how to handle those.

**Comment by greg0ire** [comment]
Retargeting to 3.1.x since 3.0.x is not maintained anymore.

**Comment by greg0ire** [comment]
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/3.0.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already push. Don't forget the `--force` option otherwise git will try to merge both things together.

**Review by greg0ire — COMMENTED** [review]
`docs/en/reference/events.rst` needs to document the new event

**Review by greg0ire — DISMISSED** [review]

**Review by SenseException — DISMISSED** [review]

**Comment by greg0ire** [comment]
Since this is not a patch, I think we should target 3.2.x :thinking:

**Comment by derrabus** [comment]
> Since this is not a patch, I think we should target 3.2.x 🤔

Yes. 3.2.x is a bit behind. I'll retarget after a merge up.

File: tests/Tests/ORM/Tools/SchemaToolTest.php
Comment (derrabus) [inline tests/Tests/ORM/Tools/SchemaToolTest.php]: Unused variable `$schema`.

File: tests/Tests/ORM/Tools/SchemaToolTest.php
Comment (derrabus) [inline tests/Tests/ORM/Tools/SchemaToolTest.php]: The test does not really show me that I was able to influence the generated schema or SQL.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: tests/Tests/ORM/Tools/SchemaToolTest.php
Comment (wmouwen) [reply tests/Tests/ORM/Tools/SchemaToolTest.php]: Tried to improve the tests. This should show you that you can access the schema which the comparator uses.

File: tests/Tests/ORM/Tools/SchemaToolTest.php
Comment (greg0ire) [inline tests/Tests/ORM/Tools/SchemaToolTest.php]:
```suggestion
            public function createTable($name): Table
```

Comment (wmouwen) [reply tests/Tests/ORM/Tools/SchemaToolTest.php]: Note that the return type of `createTable` is `\Doctrine\DBAL\Schema\Table`, as opposed to the existing import of `Doctrine\ORM\Mapping\Table`.

**Comment by wmouwen** [comment]
The inactivity was caused by doctrine/orm#11409 taking the maintainer's preference over doctrine/orm#11374. Hopefully that PR can move forward soon.

**Pull request closed without merging**
